airflow/dag_processing/processor.py

```diff
@@ -647,3 +650,17 @@ def process_file(
             self.log.exception("Error logging import errors!")
 
         return len(dagbag.dags), len(dagbag.import_errors)
+
+    def _deactivate_missing_dags(self, session: Session, dagbag: DagBag, file_path: str) -> None:
+        for dag in (
+            session.query(DagModel)
+            .filter(DagModel.fileloc == file_path)
+            .filter(DagModel.is_active)
+            .filter(~DagModel.dag_id.in_(dagbag.dag_ids))
+            .all()
+        ):
+            self.log.warning(
+                "Deactivating DAG ID %s since it no longer exists in file %s", dag.dag_id, file_path
+            )
+            dag.is_active = False
+            session.merge(dag)
```

Comment:
Performance here can be significantly improved with `UPDATE dag.is_active WHERE ...`. I’m not sure what the right SQLAlchemy syntax is for this though.

Reply:
```suggestion
    session.query(DagModel).filter(
        DagModel.fileloc == file_path,
        DagModel.is_active,
        ~DagModel.dag_id.in_(dagbag.dag_ids)
    ).update({"is_active": False})
```

Something like this should work !